### DISCUSSION TITLE
`beforeRequest` is not being applied for each virtual user in the same way

### DISCUSSION THREAD

**Discussion by bpkennedy** [artilleryio/artillery#2521] [Artillery]
Version info:

```
Artillery: 2.0.6
Node.js:   v18.18.2
OS:        darwin
```

Running this command: 
```
artillery run ./src/status-local.yml
```

I'm using a javascript function because I'll need to generate a guid before each POST request.  I have my `generateGUID` method running in the `beforeRequest` of the post.  I expected the `x` variable in my function to be incremented each time it is invoked (so for me duration and arrival that would be about 20 times) and to see something like this in the output:

```
⠹ context.vars.guid 2
context.vars.guid 3
⠸ context.vars.guid 4
context.vars.guid 5
⠼ context.vars.guid 6
⠴ context.vars.guid 7
context.vars.guid 8
⠦ context.vars.guid 9
⠧ context.vars.guid 10
context.vars.guid 11
Phase completed: unnamed (index: 0, duration: 2s) 09:49:26(-0600)

context.vars.guid 12
context.vars.guid 13
⠴ context.vars.guid 14
⠦ context.vars.guid 15
context.vars.guid 16
⠧ context.vars.guid 17
⠇ context.vars.guid 18
context.vars.guid 19
⠏ context.vars.guid 20
context.vars.guid 21
```

Instead, the output looks like this:

```
⠹ context.vars.guid 2
context.vars.guid 2
⠸ context.vars.guid 2
context.vars.guid 2
⠼ context.vars.guid 2
⠴ context.vars.guid 2
context.vars.guid 2
⠦ context.vars.guid 2
⠧ context.vars.guid 2
context.vars.guid 2
Phase completed: unnamed (index: 0, duration: 2s) 09:49:26(-0600)

context.vars.guid 3
context.vars.guid 3
⠴ context.vars.guid 3
⠦ context.vars.guid 3
context.vars.guid 3
⠧ context.vars.guid 3
⠇ context.vars.guid 3
context.vars.guid 3
⠏ context.vars.guid 3
context.vars.guid 3
```

Why would this happen and how would I achieve what I'm going for?

Files being used:
```
// ./src/functions.js
let x = 1

function generateGUID(request, context, events, next) {
  context.vars.guid = x++;
  console.log('context.vars.guid', x)
  return next();
}

module.exports = {
  generateGUID,  
}

// ./src/status-local.yml
config:
  target: 'http://localhost:3000'
  phases:
    - duration: 2
      arrivalRate: 10
  variables:
    email: 'test@example.com'
  processor: "./functions.js"
scenarios:
  - flow:
    - get:
        url: '/api/prompt'
    - think: 1
    - post:
        url: '/api/signin'
        json:
          email: '{{ email }}'
          guid: '{{ guid }}'
        capture:
          json: '$.token'
          as: 'loginToken'
        beforeRequest: generateGUID
    
```

**Comment by hassy** [comment]
hi @bpkennedy 👋 this is working as intended.

- Each VU has a separate context / its own set of variables. There's no built-in concept of shared memory between VUs in Artillery at the moment.
- You're using a module-level variable to work around it, but it's not working as intended because under the hood Artillery creates a number of worker threads to distributed the VUs across. So each of those threads has its own copy of `x` which is why you see its value incrementing but not to the value you're expecting.

If you need a unique ID for each VU, Artillery provides a `$uuid` variable - https://www.artillery.io/docs/reference/test-script#default-variables

You can also of course generate one with [`randomUUID()`](https://developer.mozilla.org/en-US/docs/Web/API/Crypto/randomUUID), or [`nanoid`](https://www.npmjs.com/package/nanoid) or some other way.

**Reply by hassy** [reply]
moving Issue to Discussions
